### DISCUSSION TITLE
Nuxt Head Next

### DISCUSSION THREAD

**Discussion by harlan-zw** [nuxt/nuxt#22632] [RFCs]
## Introduction

In v3.7.0 we introduce an experimental flag `headNext` (https://github.com/nuxt/nuxt/pull/22620) that aims to improve the DX and performance of your head.

```ts
export default defineNuxtConfig({
  experimental: {
     headNext: true
  }
})
```

We need help testing these features on a wider set of sites with the goal of enabling these by default in a future Nuxt release (v3.8.0).

If you opt-in to this experimental feature, we would love to get your feedback! Sharing your web vitals pre and post-flag would be appreciated.

## Optimizations

Currently, the improvements are scoped to two core [Unhead](https://github.com/unjs/unhead) plugins. 

## Capo.js

When rendering your head tags, we render them as they are registered, except for some critical tags.

For example, a `<title>` and `<meta viewport...>` tag will always be first in your `<head>`

 This can give you some predictability in terms of where and when you're registering your head data.

However, it never guarantees that the tags are in the best position for performance. It only guarantees that tags will be in the right position, so as not to break your site. 

In Nuxt 3.7.0, we have moved all head tags to be rendered with Unhead, giving us the ability to re-order all tags, including your apps tags alongside Nuxt core tags.

When using the `headNext` flag, [Capo.js](https://rviscomi.github.io/capo.js/) ordering will be applied. Capo.js is an awesome project trying to bring some structure to the head, making sure tags are rendered in the best order for performance.

We hope to see some performance improvements with this, especially for large sites that may have many third-party styles and scripts.

We're teaming up with the [Google Aurora](https://developer.chrome.com/blog/introducing-aurora/) team to track the performance of this new ordering. The attribute (`<html data-capo>`)  is added to make that tracking easier.

See the [CapoPlugin](https://github.com/unjs/unhead/blob/main/packages/unhead/src/optionalPlugins/capoPlugin.ts) to learn more.

**Comment by johannschopplich** [comment]
Wow, great work! Counting the days until Nuxt 3.7 release to test `experimental.headNext`. Thank you for your continued efforts regarding all things SEO in Nuxt, @harlan-zw!

**Reply by harlan-zw** [reply]
Thank you for the kind words!

**Comment by harlan-zw** [comment]
FYI the Hash Hydration Plugin will no longer be included with this flag. While it will be useful for some sites, there are certain limitations that have been discovered while testing in production that make it not worth including in the scope of this flag.

I plan to introduce some more features on this flag once 3.7 is out.

**Comment by becem-gharbi** [comment]
Hi, thanks for the great work! 

I have an issue with naiveui on this new release `v3.7` and I believe it's related to this discussion.

The existing server [plugin](https://github.com/becem-gharbi/nuxt-naiveui/blob/main/src/runtime/plugins/naive.server.ts) is used to override `ssrContext.renderMeta` by concatenating naiveui style tags to the head tag.

FYI, naiveui style tags are returned from the `collect` method, for example `<style cssr-id="_ID1_">\n_STYLE1_\n</style>\n<style cssr-id="_ID2_">\n_STYLE2_\n</style>`

Is there a way to inject these styles without overriding  `ssrContext.renderMeta`?

<hr>

**Edit:** I just have figured out how to solve this 
```js
  nuxtApp.ssrContext?.head.hooks.hook("ssr:rendered", (ctx) => {
    ctx.html.headTags += collect();
  });
```

**Reply by harlan-zw** [reply]
Sent a PR https://github.com/becem-gharbi/nuxt-naiveui/pull/17 :)

**Comment by hacknug** [comment]
After the update, `useHead` doesn't seem to be auto-imported anymore (`useHead is not defined`). Is this something expected or something that happened to others? Adding `import { useHead } from 'unhead'` solves the issue but I'm not sure what's going on since I can still see the auto-import on `imports.d.ts`

**Reply by harlan-zw** [reply]
I have not seen this before :thinking: Any reproduction :pray: ?

**Reply by hacknug** [reply]
I tried but couldn't reproduce outside of my project.

The weird thing is I can see it being exported in my `imports.d.ts` file (`export { injectHead, useHead, useSeoMeta, useHeadSafe, useServerHead, useServerSeoMeta, useServerHeadSafe } from '@unhead/vue';`).

The explicit import is also only required in `app.vue` but not on the files inside my `plugins` dir.

Value of `experimental.headNext` seems to have no effect on the outcome.

Here's the stack trace in case it hints at something I'm missing:

```
500
useHead is not defined

at _sfc_main.setup (./app.js:148:23)
at callWithErrorHandling (./node_modules/.pnpm/@vue+runtime-core@3.3.4/node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:156:18)
at setupStatefulComponent (./node_modules/.pnpm/@vue+runtime-core@3.3.4/node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:7190:25)
at setupComponent (./node_modules/.pnpm/@vue+runtime-core@3.3.4/node_modules/@vue/runtime-core/dist/runtime-core.cjs.js:7151:36)
at renderComponentVNode (./node_modules/.pnpm/@vue+server-renderer@3.3.4_vue@3.3.4/node_modules/@vue/server-renderer/dist/server-renderer.cjs.js:614:15)
at Module.ssrRenderComponent (./node_modules/.pnpm/@vue+server-renderer@3.3.4_vue@3.3.4/node_modules/@vue/server-renderer/dist/server-renderer.cjs.js:86:10)
at default (./node_modules/.pnpm/nuxt@3.7.3/node_modules/nuxt/dist/app/components/nuxt-root.js:79:37)
at Module.ssrRenderSuspense (./node_modules/.pnpm/@vue+server-renderer@3.3.4_vue@3.3.4/node_modules/@vue/server-renderer/dist/server-renderer.cjs.js:461:5)
at _sfc_ssrRender (./node_modules/.pnpm/nuxt@3.7.3/node_modules/nuxt/dist/app/components/nuxt-root.js:70:25)
```

**Reply by harlan-zw** [reply]
Here's some debugging steps:
- Have you specified any resolutions within your package.json?
- Did you upgrade with `nuxi upgrade`? 
- Run`npm why @unhead/vue`, maybe it will give a clue

**Reply by hacknug** [reply]
- No resolutions.
- Always run `nuxi upgrade --force` after updating deps.
- Just did that and npm returns a +1000 lines file I'm not sure what to do with. pnpm returns a simplified version where I can see all my deps depend on `nuxt@3.7.3` and `@unhead/vue@1.7.3` so I feel like everything looks correct.

```
devDependencies:
@histoire/plugin-nuxt 0.17.0
└─┬ nuxt 3.7.3 peer
  └── @unhead/vue 1.7.3
@nuxt/devtools 0.8.3
├─┬ @nuxt/devtools-kit 0.8.3
│ └─┬ nuxt 3.7.3 peer
│   └── @unhead/vue 1.7.3
└─┬ nuxt 3.7.3 peer
  └── @unhead/vue 1.7.3
@vueuse/nuxt 10.4.1
└─┬ nuxt 3.7.3 peer
  └── @unhead/vue 1.7.3
nuxt 3.7.3
└── @unhead/vue 1.7.3
```

I also tried on a new computer where I've never done any development on and the same error shows up so I don't think it is something related to cached or global deps, or anything similar.

It has to be related to the way I've done something on the codebase (likely since this project started with one of the first versions of nuxt 3 rc) or a conflict in my dependencies (less likely although since this issue only showed up after a nuxt update I would still consider it possible).

Unless someone else experiences it or knows what's going on, I'd say it is not worth it to look into it anymore. Manually importing `useHead` is way cheaper that debugging something that doesn't seem to be an issue in core.

Leaving this here just in case someone else experiences the issue and can add some extra context and details.

Thanks @harlan-zw ❤️

**Reply by harlan-zw** [reply]
Thanks, it does sound like something particular about your project.

If you do want a hand debugging further I'd be happy to jump into the code over Discord or something.

My only suggestion is to import from `@unhead/vue` instead of `unhead`, if you want the reactivity to work.

**Comment by kiki-kanri** [comment]
Hello, I would like to ask if I should use the [nuxt-capo](https://github.com/danielroe/nuxt-capo) module for now or just enable the headNext option in the settings.

These two approaches will have different results.

**Reply by harlan-zw** [reply]
Enable headNext, you'll get more accurate sorting 👍

**Reply by kiki-kanri** [reply]
Ok, thank you for your reply!

**Comment by Barbapapazes** [comment]
Hello,

I've enable headNext for unjs/website, see https://github.com/unjs/website/pull/254, but the head does not seems to be ordered.

![image](https://github.com/nuxt/nuxt/assets/45267552/9fff93cf-852b-47bc-bb01-6bc3cf60743c)

**Reply by harlan-zw** [reply]
This is actually due to the OG Image module inserting tag after capo.js runs

**Comment by ni-marchuk** [comment]
I used the experimental flag and also updated to the latest version of nuxt.
I previously used useSeoMeta, but rewrote it to use useHead.
The result remains the same. my meta tags remain almost at the very bottom
what should I do?
